docs(agw): Initial MME C++ migration proposal (#6987)
magma/magma: docs/readmes/proposals/p017_mme_migrate_to_c++.md

Some MM and NAS IEs, for the ease of their manipulation (comparison operators, etc) could be migrated with wrapper classes.

[A discussion of alternate approaches and the trade offs, advantages, and
disadvantages of the specified approach.]

## Compatibility

Endianness: Whatever is the HW platform, endianness is fixed.

Backward compatibility of the new serialization library with protobuf is not envisioned.

## Observability and Debug

TODO
[A description, of how issues with this design would be observed and debugged
in various stages from development through production.]

## Implementation

  Description of steps in the C++ migration:
 
* Serialization library selection
    * Build 2 realistics Prototypes with MME_APP, NAS protos (cap’n proto, Flat buffers)
        * Migrate hashtables (find functions)
    * Get measurements and evaluate performances, maintainability.
    * Select library
* MME_APP/NAS Serialization migration
    * Write proto
    * Write redis client libraries

>     * Migrate hashtables (find functions, evaluate performance)

Reviewer: @lionelgo I would move this as a separate effort. Maybe instead of hashtable migration being part of the serialization efforts, I think it should be in its own contained task.

Author: @ardzoht I change accordingly to your suggestion, It will be clearer to me when I will get more familiar with the serialization library.

Author: @ardzoht I made a commit, if I missed your point, please change task list according to your thoughts

    	Not all context attribute need to be serialized
    * Migrate states (update core code with C++ patterns)
    * Write Unit tests
* Migrate MME_APP/NAS hashtables (find functions, evaluate performance)
* S1AP Serialization migration
    * Write proto
    * Write redis client libraries
    * Design classes for wrapping generated POD classes 
    	Not all context attribute need to be serialized
    * Migrate states (update core code)
    * Write Unit tests

> * Migrate S1AP hashtables (find functions, evaluate performance)

Reviewer: nit: Can we specify all MME task threads (S1AP, MME and SPGW) state hashtables?

Author: working on it...

Author: I made an update, let me know.

Reviewer: This looks good to me, thanks!

---
id: p017_mme_migrate_to_c++
title: C++ Migration
hide_title: true
---

# Proposal: C++ Migration

> Author(s): [Author Name, Co-Author Name]

Reviewer 2: Should this have your name? :D

Author: Many items have been suggested by Ulas and Scott, I will put my name along their name also.

Last updated: 04/19/2021

Discussion at
[https://github.com/magma/magma/issues/4074](https://github.com/magma/magma/issues/4074).

## Abstract

This document describes the steps to migrate progressively, from C to C++, the MME core.

## Background

Initially and historically, the development of MME magma was done only in C.

Many third party libraries and features that were integrated and or developed around the MME core are written in C++.

We are now to the point where, for an easier and cleaner way of integrating new C++ libraries, for the ease of the maintenance of the code, a C to C++ migration seems unavoidable.

## Proposal

The proposal is constrained by a lack of performance of the current context serialization library (protobuf).
 
Because the replacement libraries may have a big impact on the MME core code, this library migration will be studied first.

### State Serialization
The serialization performance of protobuf is insufficient, so a serialization library that allows better performance must be selected.

> The bad performances are on the global serialization process (translation, memory allocation, copy).

Reviewer: ```suggestion
The bad performance is due to the global serialization process (translation, memory allocation, copy).
```

> The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standart API, binary output format and do not need memory copy (zero copy).

Reviewer: ```suggestion
The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standard API, binary output format and do not need memory copy (zero copy).
```

> The states that have to be serialized are (To be detailled):

Reviewer 3: Just this miss-spelling of detailed!

Author: Ok I will correct this very soon, thanks

Author: done

[A discussion of alternate approaches and the trade offs, advantages, and
disadvantages of the specified approach.]

## Compatibility

Endianness: Whatever is the HW platform, endianness is fixed.

Backward compatibility of the new serialization library with protobuf is not envisioned.

## Observability and Debug

TODO
[A description, of how issues with this design would be observed and debugged
in various stages from development through production.]

## Implementation

  Description of steps in the C++ migration:
 
* Serialization library selection
    * Build 2 realistics Prototypes with MME_APP, NAS protos (cap’n proto, Flat buffers)
        * Migrate hashtables (find functions)
    * Get measurements and evaluate performances, maintainability.
    * Select library
* MME_APP/NAS Serialization migration
    * Write proto
    * Write redis client libraries
    * Migrate hashtables (find functions, evaluate performance)
    * Design classes for wrapping generated POD classes (?)
    	Not all context attribute need to be serialized
    * Migrate states (update core code with C++ patterns)

>     * Write Unit tests

Reviewer 4: which parts are strictly migrating c code to c++ and which parts are refactoring of existing c++ code?

Author: I have to work a bit more to give an accurate answer.

Author: To give an answer, I may need also to work on serialization libraries, I need to sync with you and/or people actually evaluating these libraries.

## Abstract

This document describes the steps to migrate progressively, from C to C++, the MME core.

## Background

Initially and historically, the development of MME magma was done only in C.

Many third party libraries and features that were integrated and or developed around the MME core are written in C++.

We are now to the point where, for an easier and cleaner way of integrating new C++ libraries, for the ease of the maintenance of the code, a C to C++ migration seems unavoidable.

## Proposal

The proposal is constrained by a lack of performance of the current context serialization library (protobuf).
 
Because the replacement libraries may have a big impact on the MME core code, this library migration will be studied first.

### State Serialization
The serialization performance of protobuf is insufficient, so a serialization library that allows better performance must be selected.

The bad performance is due to the global serialization process (translation, memory allocation, copy).

The replacement libraries considered are "cap'n proto" and "Flat buffers”, they support C/C++ standard API, binary output format and do not need memory copy (zero copy).

> The states that have to be serialized are (To be detailed):

Reviewer 4: to be detailed in another doc or in this one?

- S1AP, S1AP_IMSI_MAP, S1AP_UE
- MME_UE, MME_NAS
- SGW, SGW_UE, SPGW, SPGW_UE
- S11/GTPV2-C
- NGAP, NGAP_IMSI_MAP, NGAP_UE

#### S1AP, MME_APP, NAS migration

 S1AP seems to have no state dependency with other tasks.
 
 Should be an opportunity to start doing a first migration with S1AP.

 MME_APP and NAS tasks have contexts in common.
 

////### Migrate some MM and NAS IEs

> ////For the ease of manipulating IEs, migrate a selected list of IEs (operator ==, copy, conversion, etc)

Reviewer 4: any plans to add this part or else let's remove it.

A C++ port of S6a interface is already available in OAI HSS repo (https://github.com/OPENAIRINTERFACE/openair-hss) could serve as a basis.

S11 Interface and its GTPV2-C protocol implemented in C should be migrated to C++.
A C++ port of S11 and GTPV2-C interface done in OAI SPGW-C repo (https://github.com/OPENAIRINTERFACE/openair-spgwc) could serve as a basis.

## Rationale

### Get/Set state attributes

For a progressive migration, and a decoupling of MME core code from the serialization library, macros, inline, helpers functions/C++ patterns may be used instead of directly using serialization library API in MME core code.

Some MM and NAS IEs, for the ease of their manipulation (comparison operators, etc) could be migrated with wrapper classes.

[A discussion of alternate approaches and the trade offs, advantages, and
disadvantages of the specified approach.]

## Compatibility

Endianness: Whatever is the HW platform, endianness is fixed.

Backward compatibility of the new serialization library with protobuf is not envisioned.

## Observability and Debug

TODO
[A description, of how issues with this design would be observed and debugged

> in various stages from development through production.]

Reviewer 4: will this section be removed or plans to add details?
